src/toil/lib/misc.py

```diff
@@ -191,3 +192,62 @@ def truncExpBackoff():
         t *= 2
     while True:
         yield t
+
+
+def atomic_tmp_file(final_path):
+    """Return a tmp file name to use with atomic_install.  This will be in the
+    same directory as final_path. The temporary file will have the same extension
+    as finalPath.  In final path is in /dev (/dev/null, /dev/stdout), it is
+    returned unchanged and atomic_tmp_install will do nothing."""
+    final_dir = os.path.dirname(os.path.normpath(final_path))
+    if final_dir == '/dev':
+        return final_path
+    if final_dir == "":
+        final_dir = '.'
+    final_basename = os.path.basename(final_path)
+    final_ext = os.path.splitext(final_path)[1]
+    base_name = "{}.{}.{}.tmp{}".format(final_basename, socket.gethostname(), os.getpid(), final_ext)
+    return  os.path.join(final_dir, base_name)
+
+
+def atomic_install(tmp_path, final_path):
+    "atomic install of tmp_path as final_path"
+    if os.path.dirname(os.path.normpath(final_path)) != '/dev':
+        os.rename(tmp_path, final_path)
+
+
+class AtomicFileCreate(object):
+    """Context manager to great an temporary file.  Entering returns path to the
+    temporary file in the same directory.  If the code in context succeeds,
+    the file renamed to it's actually nape.  If an error occurs, the file is
+    not installed and is removed unless keeps is specified.
```

Comment:
Minor typos:
- s/great/create/
- s/it's/its/
- s/nape/name/
- s/keeps/keep/

Beyond this, this patch looks solid to me!